File: ios/HybridMailMailbox.swift
//
//  HybridMailMailbox.swift
//  react-native-mail-engine
//
//  A selected IMAP mailbox. All command work runs on the account's serial queue;
//  IDLE runs on its own dedicated thread so `startIdle` resolves immediately and
//  push notifications stream in until `stopIdle`.
//

import Foundation
import NitroModules

final class HybridMailMailbox: HybridMailMailboxSpec {
  private let session: RNMailSession
  private let queue: DispatchQueue
  private let folderPath: String
  private let readOnly: Bool

  private let existsValue: Double
  private let unseenValue: Double
  private let uidNextValue: Double
  private let uidValidityValue: Double

  private var idling = false
  private var idleThread: Thread?

  init(session: RNMailSession, queue: DispatchQueue, path: String, status: RNFolderStatus, readOnly: Bool) {
    self.session = session
    self.queue = queue
    self.folderPath = path
    self.readOnly = readOnly
    self.existsValue = Double(status.messageCount)
    self.unseenValue = Double(status.unseenCount)
    self.uidNextValue = Double(status.uidNext)
    self.uidValidityValue = Double(status.uidValidity)
    super.init()
  }

  /// Throw if a mutating op is attempted on a read-only (EXAMINE-intent) mailbox.
  private func ensureWritable() throws {
    if readOnly {
      throw MailEngineError(code: "ERR_UNSUPPORTED", message: "Mailbox was opened read-only")
    }
  }

  var path: String { folderPath }
  var exists: Double { existsValue }
  var unseen: Double { unseenValue }
  var uidNext: Double { uidNextValue }
  var uidValidity: Double { uidValidityValue }

  // MARK: - Fetch

  func fetchHeaders(options: MailFetchHeadersOptions) throws -> Promise<[MailHeaderStruct]> {
    let session = self.session
    let folderPath = self.folderPath
    let limit = options.limit.map { Int($0) }
    let sinceUid = options.sinceUid.map { UInt64($0) }
    let uidLo = options.uidRangeStart.map { UInt64($0) }
    let uidHi = options.uidRangeEnd.map { UInt64($0) }
    let sinceMs = options.sinceDateMs
    let beforeMs = options.beforeDateMs

    return runMail(on: queue) { () -> [MailHeaderStruct] in
      // Pick candidate UIDs: a server-side search when date-filtered, else all.
      var uids: [NSNumber]
      if sinceMs != nil || beforeMs != nil {
        let crit = RNSearchCriteria()
        if let s = sinceMs { crit.sinceDateMs = NSNumber(value: s) }
        if let b = beforeMs { crit.beforeDateMs = NSNumber(value: b) }
        uids = try session.search(inFolder: folderPath, criteria: crit)
      } else {
        uids = try session.allUids(inFolder: folderPath)
      }

      uids = uids.filter { n in
        let v = n.uint64Value
        if let su = sinceUid, v <= su { return false }
        if let lo = uidLo, v < lo { return false }
        if let hi = uidHi, v > hi { return false }
        return true
      }
      uids.sort { $0.uint64Value > $1.uint64Value } // newest first
      if let limit = limit, uids.count > limit {
        uids = Array(uids.prefix(limit))
      }
      if uids.isEmpty { return [] }

      return try session.fetchHeaders(inFolder: folderPath, uids: uids).map(toStruct)
    }
  }

  func fetchMessage(uid: Double, options: MailFetchMessageOptions) throws -> Promise<MailMessageStruct> {
    let session = self.session
    let folderPath = self.folderPath
    let include = options.includeAttachments ?? true
    let maxBytes = Int64(options.maxAttachmentBytes ?? 0)
    let markSeen = (options.markSeen ?? false) && !readOnly
    let messageUid = UInt32(uid)

    return runMail(on: queue) {
      let message = try session.fetchMessage(
        inFolder: folderPath,
        uid: messageUid,
        includeAttachments: include,
        maxAttachmentBytes: maxBytes,
        markSeen: markSeen
      )
      return toStruct(message)
    }
  }

  func fetchAttachment(uid: Double, partId: String) throws -> Promise<ArrayBuffer> {
    let session = self.session
    let folderPath = self.folderPath
    let messageUid = UInt32(uid)
    return runMail(on: queue) {
      let data = try session.fetchAttachment(inFolder: folderPath, uid: messageUid, partId: partId)
      return try ArrayBuffer.copy(data: data)
    }
  }

  func search(criteria: MailSearchCriteria) throws -> Promise<[Double]> {
    let session = self.session
    let folderPath = self.folderPath
    let rnCriteria = toRNSearch(criteria)
    return runMail(on: queue) {
      try session.search(inFolder: folderPath, criteria: rnCriteria).map { $0.doubleValue }
    }
  }

  // MARK: - Flags / move / delete

  private func uidsToNumbers(_ uids: [Double]) -> [NSNumber] {
    return uids.map { NSNumber(value: $0) }
  }

  func addFlags(uids: [Double], flags: [String]) throws -> Promise<Void> {
    let session = self.session
    let folderPath = self.folderPath
    let nums = uidsToNumbers(uids)
    return runMail(on: queue) { try self.ensureWritable(); try session.storeFlags(inFolder: folderPath, uids: nums, flags: flags, mode: 0) }
  }

  func removeFlags(uids: [Double], flags: [String]) throws -> Promise<Void> {
    let session = self.session
    let folderPath = self.folderPath
    let nums = uidsToNumbers(uids)
    return runMail(on: queue) { try self.ensureWritable(); try session.storeFlags(inFolder: folderPath, uids: nums, flags: flags, mode: 1) }
  }

  func markSeen(uids: [Double], seen: Bool) throws -> Promise<Void> {
    let session = self.session
    let folderPath = self.folderPath
    let nums = uidsToNumbers(uids)
    return runMail(on: queue) {
      try self.ensureWritable()
      try session.storeFlags(inFolder: folderPath, uids: nums, flags: ["\\Seen"], mode: seen ? 0 : 1)
    }
  }

  func moveMessages(uids: [Double], destinationPath: String) throws -> Promise<Void> {
    let session = self.session
    let folderPath = self.folderPath
    let nums = uidsToNumbers(uids)
    return runMail(on: queue) { try self.ensureWritable(); try session.moveMessages(inFolder: folderPath, uids: nums, destination: destinationPath) }
  }

  func copyMessages(uids: [Double], destinationPath: String) throws -> Promise<Void> {
    let session = self.session
    let folderPath = self.folderPath
    let nums = uidsToNumbers(uids)
    return runMail(on: queue) { try session.copyMessages(inFolder: folderPath, uids: nums, destination: destinationPath) }
  }

  func deleteMessages(uids: [Double], expunge: Bool) throws -> Promise<Void> {
    let session = self.session
    let folderPath = self.folderPath
    let nums = uidsToNumbers(uids)
    return runMail(on: queue) { try self.ensureWritable(); try session.deleteMessages(inFolder: folderPath, uids: nums, expunge: expunge) }
  }

  // MARK: - IDLE

  func startIdle(
    onMail: @escaping (MailNewMailEvent) -> Void,
    onError: @escaping (MailErrorStruct) -> Void
  ) throws -> Promise<Void> {
    let session = self.session
    let folderPath = self.folderPath
    guard session.serverSupportsIdle() else {
      return Promise.rejected(
        withError: MailEngineError(code: "ERR_UNSUPPORTED", message: "Server does not support IMAP IDLE")
      )
    }
    if idling { return Promise.resolved() }
    idling = true
    var lastUid = UInt32(uidNextValue > 0 ? uidNextValue - 1 : 0)

    let thread = Thread { [weak self] in
      while self?.idling == true {
        do {
          try session.idleOnce(inFolder: folderPath, lastKnownUid: lastUid) { uids, exists in
            if let maxUid = uids.map({ $0.uint32Value }).max() { lastUid = maxUid }
            onMail(MailNewMailEvent(uids: uids.map { $0.doubleValue }, exists: Double(exists)))
          }
        } catch let error as NSError {
          let mapped = bridgeError(error)
          onError(MailErrorStruct(code: mapped.code, message: mapped.message))
          break
        }
      }
    }
    thread.stackSize = 1 << 19
    idleThread = thread
    thread.start()
    return Promise.resolved()
  }

  func stopIdle() throws -> Promise<Void> {
    idling = false
    session.interruptIdle()
    idleThread = nil
    return Promise.resolved()
  }

  func close() throws -> Promise<Void> {
    idling = false
    session.interruptIdle()
    return Promise.resolved()
  }
}

File: android/src/main/java/com/margelo/nitro/mailengine/HybridMailMailbox.kt
package com.margelo.nitro.mailengine

import com.margelo.nitro.core.ArrayBuffer
import com.margelo.nitro.core.Promise
import com.sun.mail.imap.IMAPFolder
import java.util.Date
import java.util.concurrent.ExecutorService
import javax.mail.FetchProfile
import javax.mail.Flags
import javax.mail.Folder
import javax.mail.Message
import javax.mail.UIDFolder
import javax.mail.event.MessageCountAdapter
import javax.mail.event.MessageCountEvent
import javax.mail.search.AndTerm
import javax.mail.search.BodyTerm
import javax.mail.search.ComparisonTerm
import javax.mail.search.FlagTerm
import javax.mail.search.FromStringTerm
import javax.mail.search.OrTerm
import javax.mail.search.ReceivedDateTerm
import javax.mail.search.RecipientStringTerm
import javax.mail.search.SearchTerm
import javax.mail.search.SubjectTerm

/**
 * A selected IMAP mailbox. Command work runs on the account's serial [executor].
 * IDLE runs on its own daemon thread; it is broken by issuing a folder command
 * from the executor thread (`folder.getMessageCount()`), the documented JavaMail
 * way to terminate a blocked `idle()`.
 */
class HybridMailMailbox(
  private val folder: IMAPFolder,
  private val executor: ExecutorService
) : HybridMailMailboxSpec() {

  // Snapshot at open time — Nitro property reads happen on the JS thread, which
  // must never touch the (non-thread-safe) folder concurrently with the executor.
  private val pathValue = folder.fullName
  private val existsValue = runCatching { folder.messageCount }.getOrDefault(0).toDouble()
  private val unseenValue = runCatching { folder.unreadMessageCount }.getOrDefault(0).toDouble()
  private val uidNextValue = runCatching { folder.uidNext }.getOrDefault(0L).toDouble()
  private val uidValidityValue = runCatching { folder.uidValidity }.getOrDefault(0L).toDouble()

  @Volatile private var idling = false
  private var idleThread: Thread? = null

  override val path: String get() = pathValue
  override val exists: Double get() = existsValue
  override val unseen: Double get() = unseenValue
  override val uidNext: Double get() = uidNextValue
  override val uidValidity: Double get() = uidValidityValue

  // MARK: - Fetch

  override fun fetchHeaders(options: MailFetchHeadersOptions): Promise<Array<MailHeaderStruct>> =
    MailBridge.run(executor) {
      // Narrow the fetch to a server-side UID range so a large mailbox doesn't
      // pull every envelope just to return the newest few.
      val lo = maxOf(
        options.sinceUid?.let { it.toLong() + 1 } ?: 1L,
        options.uidRangeStart?.toLong() ?: 1L
      )
      val hi = options.uidRangeEnd?.toLong() ?: UIDFolder.LASTUID
      val messages = folder.getMessagesByUID(lo, hi)
      val profile = FetchProfile().apply {
        add(FetchProfile.Item.ENVELOPE)
        add(FetchProfile.Item.FLAGS)
        add(FetchProfile.Item.CONTENT_INFO)
        add(UIDFolder.FetchProfileItem.UID)
        add(IMAPFolder.FetchProfileItem.SIZE)
      }
      folder.fetch(messages, profile)

      var entries = messages.map { it to folder.getUID(it) }
      options.sinceUid?.let { su -> entries = entries.filter { it.second > su.toLong() } }
      options.uidRangeStart?.let { lo -> entries = entries.filter { it.second >= lo.toLong() } }
      options.uidRangeEnd?.let { hi -> entries = entries.filter { it.second <= hi.toLong() } }
      options.sinceDateMs?.let { s ->
        entries = entries.filter { ((it.first.receivedDate ?: it.first.sentDate)?.time ?: 0L) >= s.toLong() }
      }
      options.beforeDateMs?.let { b ->
        entries = entries.filter { ((it.first.receivedDate ?: it.first.sentDate)?.time ?: Long.MAX_VALUE) < b.toLong() }
      }
      entries = entries.sortedByDescending { it.second }
      options.limit?.let { lim -> entries = entries.take(lim.toInt()) }

      entries.map { MailBridge.headerOf(it.first, it.second) }.toTypedArray()
    }

  override fun fetchMessage(uid: Double, options: MailFetchMessageOptions): Promise<MailMessageStruct> =
    MailBridge.run(executor) {
      val message = folder.getMessageByUID(uid.toLong())
        ?: throw MailCodedException("ERR_PARSE", "Message with UID $uid not found")
      val parsed = MailBridge.parseMessage(
        message,
        uid.toLong(),
        options.includeAttachments ?: true,
        (options.maxAttachmentBytes ?: 0.0).toLong()
      )
      if (options.markSeen == true) {
        runCatching { message.setFlag(Flags.Flag.SEEN, true) }
      }
      parsed
    }

  override fun fetchAttachment(uid: Double, partId: String): Promise<ArrayBuffer> =
    MailBridge.run(executor) {
      val message = folder.getMessageByUID(uid.toLong())
        ?: throw MailCodedException("ERR_PARSE", "Message with UID $uid not found")
      val parsed = MailBridge.parseMessage(message, uid.toLong(), includeAttachments = true, maxBytes = 0L)
      parsed.attachments.firstOrNull { it.partId == partId }?.data
        ?: throw MailCodedException("ERR_PARSE", "Attachment part $partId not found")
    }

  override fun search(criteria: MailSearchCriteria): Promise<DoubleArray> = MailBridge.run(executor) {
    val term = buildSearchTerm(criteria)
    val messages = if (term != null) folder.search(term) else folder.messages
    var uids = messages.map { folder.getUID(it) }
    criteria.uidRangeStart?.let { lo -> uids = uids.filter { it >= lo.toLong() } }
    criteria.uidRangeEnd?.let { hi -> uids = uids.filter { it <= hi.toLong() } }
    uids.map { it.toDouble() }.toDoubleArray()
  }

  // MARK: - Flags / move / delete

  private fun messagesByUids(uids: DoubleArray): Array<Message> =
    uids.mapNotNull { folder.getMessageByUID(it.toLong()) }.toTypedArray()

  override fun addFlags(uids: DoubleArray, flags: Array<String>): Promise<Unit> = MailBridge.run(executor) {
    val messages = messagesByUids(uids)
    if (messages.isNotEmpty()) folder.setFlags(messages, MailBridge.toFlags(flags), true)
    Unit
  }

  override fun removeFlags(uids: DoubleArray, flags: Array<String>): Promise<Unit> = MailBridge.run(executor) {
    val messages = messagesByUids(uids)
    if (messages.isNotEmpty()) folder.setFlags(messages, MailBridge.toFlags(flags), false)
    Unit
  }

  override fun markSeen(uids: DoubleArray, seen: Boolean): Promise<Unit> = MailBridge.run(executor) {
    val messages = messagesByUids(uids)
    if (messages.isNotEmpty()) folder.setFlags(messages, Flags(Flags.Flag.SEEN), seen)
    Unit
  }

  override fun moveMessages(uids: DoubleArray, destinationPath: String): Promise<Unit> = MailBridge.run(executor) {
    val messages = messagesByUids(uids)
    if (messages.isNotEmpty()) {
      val destination = folder.store.getFolder(destinationPath)
      folder.copyMessages(messages, destination)
      folder.setFlags(messages, Flags(Flags.Flag.DELETED), true)
      folder.expunge()
    }
    Unit
  }

  override fun copyMessages(uids: DoubleArray, destinationPath: String): Promise<Unit> = MailBridge.run(executor) {
    val messages = messagesByUids(uids)
    if (messages.isNotEmpty()) {
      folder.copyMessages(messages, folder.store.getFolder(destinationPath))
    }
    Unit
  }

  override fun deleteMessages(uids: DoubleArray, expunge: Boolean): Promise<Unit> = MailBridge.run(executor) {
    val messages = messagesByUids(uids)
    if (messages.isNotEmpty()) {
      folder.setFlags(messages, Flags(Flags.Flag.DELETED), true)
      if (expunge) folder.expunge()
    }
    Unit
  }

  private fun buildSearchTerm(c: MailSearchCriteria): SearchTerm? {
    val terms = ArrayList<SearchTerm>()
    c.text?.let { terms.add(OrTerm(arrayOf(SubjectTerm(it), BodyTerm(it), FromStringTerm(it)))) }
    c.from?.let { terms.add(FromStringTerm(it)) }
    c.to?.let { terms.add(RecipientStringTerm(Message.RecipientType.TO, it)) }
    c.subject?.let { terms.add(SubjectTerm(it)) }
    c.body?.let { terms.add(BodyTerm(it)) }
    c.seen?.let { terms.add(FlagTerm(Flags(Flags.Flag.SEEN), it)) }
    c.flagged?.let { terms.add(FlagTerm(Flags(Flags.Flag.FLAGGED), it)) }
    c.answered?.let { terms.add(FlagTerm(Flags(Flags.Flag.ANSWERED), it)) }
    c.sinceDateMs?.let { terms.add(ReceivedDateTerm(ComparisonTerm.GE, Date(it.toLong()))) }
    c.beforeDateMs?.let { terms.add(ReceivedDateTerm(ComparisonTerm.LT, Date(it.toLong()))) }
    return when {
      terms.isEmpty() -> null
      terms.size == 1 -> terms[0]
      else -> AndTerm(terms.toTypedArray())
    }
  }

  // MARK: - IDLE

  override fun startIdle(
    onMail: (event: MailNewMailEvent) -> Unit,
    onError: (error: MailErrorStruct) -> Unit
  ): Promise<Unit> = MailBridge.run(executor) {
    if (idling) return@run Unit
    idling = true
    var lastUid = runCatching { folder.uidNext - 1 }.getOrDefault(0L)

    folder.addMessageCountListener(object : MessageCountAdapter() {
      override fun messagesAdded(event: MessageCountEvent) {
        try {
          val newUids = event.messages
            .map { runCatching { folder.getUID(it) }.getOrDefault(0L) }
            .filter { it > lastUid }
          if (newUids.isNotEmpty()) lastUid = newUids.max()
          val exists = runCatching { folder.messageCount }.getOrDefault(0).toDouble()
          onMail(MailNewMailEvent(newUids.map { it.toDouble() }.toDoubleArray(), exists))
        } catch (t: Throwable) {
          onError(MailErrorStruct("ERR_IMAP", t.message ?: "IDLE notification error"))
        }
      }
    })

    val thread = Thread({
      while (idling) {
        try {
          folder.idle()
        } catch (t: Throwable) {
          if (idling) {
            val coded = MailBridge.coded(t) as? MailCodedException
            onError(MailErrorStruct(coded?.code ?: "ERR_IMAP", t.message ?: "IDLE connection dropped"))
          }
          break
        }
      }
    }, "rn-mail-engine-idle-$pathValue").apply { isDaemon = true }
    idleThread = thread
    thread.start()
    Unit
  }

  override fun stopIdle(): Promise<Unit> = MailBridge.run(executor) {
    idling = false
    // Issue a command from the executor thread to break the blocked idle().
    runCatching { folder.messageCount }
    idleThread = null
    Unit
  }

  override fun close(): Promise<Unit> = MailBridge.run(executor) {
    idling = false
    runCatching { if (folder.isOpen) folder.close(false) }
    Unit
  }
}
